File: 7 kyu/The First Non Repeated Character In A String.js
/*
https://www.codewars.com/kata/570f6436b29c708a32000826

You need to write a function, that returns the first non-repeated character in the given string.

If all the characters are unique, return the first character of the string.
If there is no unique character, return null in JS or Java, and None in Python.

You can assume, that the input string has always non-zero length.

Example
firstNonRepeated("test") // returns "e"
firstNonRepeated("teeter") // returns "r"
firstNonRepeated("trend") // returns "t" (all the characters are unique)
firstNonRepeated("aabbcc") // returns null (all the characters are repeated)
*/

// my solution
function firstNonRepeated(s) {
  let len = s.length;
  let uniLen = new Set(s).size;
  if (uniLen === len) return s[0];
  if (uniLen === 0) return null;

  let result = [...s].find((x) => s.indexOf(x) === s.lastIndexOf(x));
  return result === undefined ? null : result;
}

//! alternative-solution
const firstNonRepeated = (s) =>
  [...s].find((i) => s.indexOf(i) == s.lastIndexOf(i)) || null;
// should have used that  || null trick for my solution too

File: Java/7kyu/The First Non Repeated Character In A String.java
/*
https://www.codewars.com/kata/570f6436b29c708a32000826

You need to write a function, that returns the first non-repeated character in the given string.

If all the characters are unique, return the first character of the string.
If there is no unique character, return null in JS or Java, and None in Python.

You can assume, that the input string has always non-zero length.

Example
firstNonRepeated("test") // returns "e"
firstNonRepeated("teeter") // returns "r"
firstNonRepeated("trend") // returns "t" (all the characters are unique)
firstNonRepeated("aabbcc") // returns null (all the characters are repeated)
*/

// my solution
import java.util.Map;
import java.util.LinkedHashMap;

public class FirstNonRepeated {
    public static Character firstNonRepeated(String source) {
        LinkedHashMap<Character, Integer> dict = new LinkedHashMap<>();

        for (char c : source.toCharArray()) {
            dict.put(c, dict.getOrDefault(c, 0) + 1);
        }

        for (Map.Entry<Character, Integer> item : dict.entrySet()) {
            if (item.getValue() == 1) {
                return item.getKey();
            }
        }
        
        return null;
    }
}